clients/vscode-hlasmplugin/src/hlasmDownloadCommands.ts

```diff
@@ -0,0 +1,697 @@
+        return {
+            unpacker: Promise.reject(e)
+        };
+    }
+}
+
+export async function downloadDependenciesWithClient(client: JobClient,
+    jobList: JobDetail[],
+    jobcardPattern: string,
+    progress: StageProgressReporter,
+    io: IoOps,
+    cancelled: () => boolean): Promise<{ failed: JobDetail[]; total: number; }> {
+
+    const checkCancel = () => {
+        if (cancelled())
+            throw Error("Cancel requested");
+    };
+
+    try {
+        const jobcard = prepareJobHeader(jobcardPattern);
+        const jobs = await submitJobs(client, jobcard, jobList, progress, checkCancel);
+        const jobsMap: {
+            [key: string]: SubmittedJob
+        } = Object.assign({}, ...jobs.map(x => ({ [x.jobname + "." + x.jobid]: x })));
+
+        await client.setListMask(jobcard.jobMask);
+
+        let wait = 0;
+        let result = { failed: new Array<JobDetail>(), total: 0 };
+        while (jobs.some(x => !x.downloaded)) {
+            checkCancel();
+
+            const list = (await client.list()).map(x => {
+                const j = jobsMap[x.jobname + "." + x.id];
+                return { fileInfo: x, job: j && !j.downloaded ? j : null };
+            }).filter(x => !!x.job);
+
+            for (const l of list) {
+                const job = l.job!;
+                job.unpacking = (await downloadJobAndProcess(client, l.fileInfo, job, progress, io)).unpacker
+                    .then(_ => { result.total++; })
+                    .catch(_ => { result.total++; result.failed.push(job.details); });
+            }
+
+            if (list.length === 0) {
+                if (wait < 30)
+                    wait += 1;
+                await new Promise((resolve) => setTimeout(resolve, wait * 1000))
+            }
+            else
+                wait = 0;
+        }
+
+        await Promise.all(jobs.map(x => x.unpacking));
+
+        return result;
+    }
+    finally {
+        client.close();
+    }
+}
+
+function askUser(prompt: string, password: boolean, defaultValue: string = ''): Promise<string> {
+    const input = vscode.window.createInputBox();
+    return new Promise<string>((resolve, reject) => {
+        input.prompt = prompt;
+        input.password = password;
+        input.value = defaultValue || '';
+        input.onDidHide(() => reject("Action was cancelled"));
+        input.onDidAccept(() => resolve(input.value));
+        input.show();
+    }).finally(() => { input.dispose(); });
+}
+
+function pickUser<T>(title: string, options: { label: string, value: T }[]): Promise<T> {
+    const input = vscode.window.createQuickPick();
+    return new Promise<T>((resolve, reject) => {
+        input.title = title;
+        input.items = options.map(x => { return { label: x.label }; });
+        input.canSelectMany = false;
+        input.onDidHide(() => reject("Action was cancelled"));
+        input.onDidAccept(() => resolve(options.find(x => x.label === input.selectedItems[0].label)!.value));
```

Comment:
The null assertion (!) is unnecessary. Typescript should properly infer the type.

Reply:
until you turn on strict checking...